packages/client/src/runtime/core/protocol/json/serialize.ts

```diff
@@ -72,13 +99,17 @@ function serializeSelectionSet(
   context: SerializeContext,
 ): JsonSelectionSet {
   if (select && include) {
-    throw new Error('select and include are used at the same time')
+    context.throwValidationError({ kind: 'IncludeAndSelect', selectionPath: context.getSelectionPath() })
```

Comment:
I wondered if this could be done in the engines rather?

Reply:
This and `IncludeOnScalar` can not - `include` and `select` exist only on the client, on protocol/engine level it is just a single selection, so engine can't tell that both `include` and `select` were used. These 2 are the only validation rules left on the client though.